linker/shared/src/main/scala/org/scalajs/linker/backend/wasmemitter/Preprocessor.scala

```diff
@@ -0,0 +1,458 @@
+     * type.
+     *
+     * Assign a bucket to a spine type if adding it does not violate the bucket
+     * assignment rule, namely: two spine types can share a bucket only if they
+     * do not have any common join type descendants. If no existing bucket
+     * satisfies the rule, create a new bucket.
+     *
+     * **2. Assign buckets to non-spine types (plain and join types)**
+     *
+     * Visit these types in level order (from root to leaves). For each type,
+     * compute the set of buckets already used by its ancestors. Assign the
+     * type to any available bucket not in this set. If no available bucket
+     * exists, create a new one.
+     */
+
+    def getAllInterfaces(clazz: LinkedClass): List[ClassName] =
+      clazz.ancestors.filter(isInterface)
+
+    val buckets = new mutable.ListBuffer[Bucket]()
+
+    def findOrCreateBucketSuchThat(p: Bucket => Boolean): Bucket = {
+      buckets.find(p).getOrElse {
+        val newBucket = new Bucket()
+        buckets += newBucket
+        newBucket
+      }
+    }
+
+    /** All join type descendants of the class */
+    val joinsOf =
+      new mutable.HashMap[ClassName, mutable.HashSet[ClassName]]()
+
+    /** the buckets that have been assigned to any of the ancestors of the class */
+    val usedOf = new mutable.HashMap[ClassName, mutable.HashSet[Bucket]]()
+    val spines = new mutable.HashSet[ClassName]()
+
+    // Phase 1: Assign buckets to spine types
+    for (clazz <- classes.reverseIterator) {
+      val ifaces = getAllInterfaces(clazz)
+
+      if (ifaces.nonEmpty) {
+        val joins = joinsOf.getOrElse(clazz.name.name, new mutable.HashSet())
+
+        if (joins.nonEmpty) {
+          // This type is a spine type
+
+          /* Look for an existing bucket to add the spine type to.
+           * Two spine types can share a bucket only if they don't have any
+           * common join type descendants.
+           */
+          val bucket = findOrCreateBucketSuchThat(!_.joins.exists(joins))
+          bucket.add(clazz.name.name)
+          bucket.joins ++= joins
+
+          for (iface <- ifaces)
+            joinsOf.getOrElseUpdate(iface, new mutable.HashSet()) ++= joins
+          spines.add(clazz.name.name)
+        } else if (ifaces.length > 1) {
+          // This type is a join type: add to joins map, bucket assignment is done later
+          for (iface <- ifaces)
+            joinsOf.getOrElseUpdate(iface, new mutable.HashSet()) += clazz.name.name
+        } else {
+          // This type is a plain type. Do nothing.
+        }
+      }
+    }
+
+    // Phase 2: Assign buckets to non-spine types (plain and join types)
+    for (clazz <- classes) {
+      val ifaces = getAllInterfaces(clazz)
+
+      if (ifaces.nonEmpty && !spines.contains(clazz.name.name)) {
+        val used = usedOf.getOrElse(clazz.name.name, new mutable.HashSet())
+        for {
+          iface <- ifaces
+          parentUsed <- usedOf.get(iface)
```

Comment:
Besides `usedOf` never being mutated, IIUC this will not contain spine types? Is that OK? If yes, comment?

Also, doing this for all and not just direct ancestors feels odd. Direct ancestors already should have transitively included buckets of their ancestors, no?